Hooks

saitonakamura opened the pull request:
Bring an api alternative via hooks
<!--- Provide a general summary of your changes in the Title above -->

## Description
There are `useIntersectionObserver` base hook
<!--- Describe your changes in detail -->

## Motivation and context
<!--- Why is this change required? What problem does it solve? -->
Fixes #130 

## How has this been tested?
Unit tests and storybook
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Screenshots (if relevant):

## Type of change
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project
- [x] My change requires a change to the documentation
- [x] I have updated the documentation accordingly
- [x] I have read the [CONTRIBUTING](CONTRIBUTING.md) document
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed

Rendez commented:
Amazing work! Sorry about the delay but I was working on the TS branch for quite long and I just finally released it. Would you mind converting your hooks to TS? (if the tests stay in JS, it's fine).

saitonakamura commented:
@Rendez funny that continued the work on it almost the same time you wrote a comment

I've migrated those to ts, I have also extracted common types into `types.ts` to be reused between existing component and a hook and improved the typings of `observer.ts` (hope you don't mind my changes)

Rendez commented:
That's perfect actually. Give me a bit of time next week to review this propertly, but it's looking good. Thanks again!

Rendez commented:
Oh, and feel free to add yourself to AUTHORS

saitonakamura commented:
@Rendez I've added disabled option support and added myself to authors and cleanup up a bit, PR should be good from now

I'd be waiting for a review

saitonakamura commented:
Coverage fails are probably coming from here https://github.com/researchgate/react-intersection-observer/commit/9ffe804e59b5af430350113ba5c1d8a81ebc585d, but i'm not sure

Rendez commented:
A quick look over the hook code showed that you return [ref, unobserve]. Why the callback in the return? Normally it's used within `onChange`

saitonakamura commented:
To let it be called from other places than onChange handler. Say I want to stop observing after a user interaction, button click for instance
With this it's gonna look like
```js
const App = () => {
  const handleChange = (e) => {
    // do stuff
  }
  const [ref, unobserve] = useIntersectionObserver(handleChange)
 return <>
    <button onClick={unobserve}>Stop</button>
    <div ref={ref}>blabla</div>
  </> 
}
```

Without it

```js
const App = () => {
  const unobserveRef = useRef()
  const handleChange = (e) => {
    unobserveRef.current = unobserve
    // do stuff
  }
  const [ref] = useIntersectionObserver(handleChange)
 return <>
    <button onClick={unobserveRef.current}>Stop</button>
    <div ref={ref}>blabla</div>
  </> 
}
```

Rendez commented:
I get that and find it very useful. With the class implementation you can accomplish the same (and should) setting the prop `disabled` to `true`. To your case it saves the user from setting state in the rendering component to pass disabled down to the hook arguments/class prop, so it's a good thing. I would still like to have in as a second argument in `onChange`, because very often it's needed to replicate a "observe only until visible" behavior. For example, we track whether an item was viewed and don't need keep observing it anymore, so it's perfect for disabling.

saitonakamura commented:
Well, I _do_ have the second argument for handleChange
https://github.com/researchgate/react-intersection-observer/pull/131/files#diff-87d18c63a0670ff32f8c78d495bc1844R7
Otherwise I was unable to implement OnlyOnce scenario

saitonakamura commented:
Hey @Rendez how's it going? Don't want to be pushy, but if you can find some time to look at it, it'd be great 🙂

Rendez commented:
@saitonakamura have you seen the comments in code? There're 3 reviews opened.

saitonakamura commented:
@Rendez no i don't see any comments 🤔  Maybe it's still in the draft or something?

Rendez commented:
Still don't see them? 
![image](https://user-images.githubusercontent.com/61776/85010185-dd3d1a00-b15f-11ea-9a11-777b9ac21f76.png)

saitonakamura commented:
@Rendez sorry I've been busy last 2 weeks, but i'll into this asap

saitonakamura commented on src/hook.ts:
```diff
@@ -0,0 +1,102 @@
+  const observingRef = useRef(false);
+
+  const instanceRef = useRef<Instance>({
+    handleChange: (ev) => onChange(ev, () => {}),
```
Yeah, I too find it, uhm, not ideal, to put it mildly, but the issue is, `unobserve` function needs an `instance` and `instance.handleChange` needs an `unobserve` to be caught by closure. So it's essentially a circular reference resolved by assigning it here https://github.com/researchgate/react-intersection-observer/pull/131/files#diff-298fdbe874b8414acb60ef615c102ec8R53

I'm gonna leave a comment about, and you have an idea how we can do this better, let me know

saitonakamura commented on src/hook.ts:
```diff
@@ -0,0 +1,102 @@
+    (node) => {
+      const isNewNode = node != null && instanceRef.current.target !== node;
+
+      if (!observer) {
```
If the observer is disabled, useMemo will recalculate and return undefined
https://github.com/researchgate/react-intersection-observer/pull/131/files#diff-298fdbe874b8414acb60ef615c102ec8R59

I also thought about depending on `disabled` prop and checking it here, but we still need to depend on `observer` and other questions arise: what will be in the `observer` if it's disabled from the start? If null/undefined we'd still need to check for it, therefore duplicating the check. If not, how to create a dummy observer?

The current way is that setRef logic only depends on availability of the observer

saitonakamura commented on src/hook.ts:
```diff
@@ -0,0 +1,102 @@
+    observe();
+
+    return obs;
+  }, [root, rootMargin, threshold, disabled]);
```
I memoized threshold with this approach https://github.com/researchgate/react-intersection-observer/pull/131/files#diff-298fdbe874b8414acb60ef615c102ec8R7 Let me know if you see any problems with it

saitonakamura commented:
@Rendez I've gone through review done some fixes and clarification, thanks for looking at this!

saitonakamura commented:
@Rendez if you some time to look at it, i'd be grateful 🙂

Rendez commented:
I'm figuring out the storybook problems and will merge after that.

saitonakamura commented:
Great, thank you!

saitonakamura commented:
🎉 Thanks for your work once again
